### PR TITLE
[docs] outline sharded ddp doc

### PR DESCRIPTION
**Pull request by stas00**
This PR provides an initial outline of HF Trainer integration starting with ZeRO. We have fairscale's Sharded optimizer/gradients supported already and deepspeed is coming 

We won't merge this until fairscale merged all the required fixes and released a new version, but I thought it'd be good to start the doc going so it's ready when fairscale is ready.

I hope to submit a deepspeed integration shortly as well, so we will extend it then with deepspeed info. edit (https://github.com/huggingface/transformers/pull/9211)

@sgugger

### CODE REVIEW

File: docs/source/training.rst [hunk 1]
```diff
@@ -278,6 +278,40 @@ pass it to the trainer.
...
 
+Trainer Integrations
+-----------------------------------------------------------------------------------------------------------------------
+
+The trainer is being extended to support experimental libraries that may dramatically improve your training time and
+fit bigger models.
+
+The main part that is being integrated at the moment is based on the paper `ZeRO: Memory Optimizations Toward Training
+Trillion Parameter Models, by Samyam Rajbhandari, Jeff Rasley, Olatunji Ruwase, Yuxiong He
+<https://arxiv.org/abs/1910.02054>`_.
```
Comment (sgugger) [inline docs/source/training.rst:289]:
```suggestion
<https://arxiv.org/abs/1910.02054>`__.
```

File: docs/source/training.rst [hunk 2]
```diff
@@ -278,6 +278,40 @@ pass it to the trainer.
...
+
+You can already deploy the following features from this paper:
+
+* Optimizer State Sharding
+* Gradient Sharding
+
+using the `--sharded_ddp` trainer argument. This is implemented via `fairscale
+<https://github.com/facebookresearch/fairscale/>`_. You will have to install `fairscale`.
```
Comment (sgugger) [inline docs/source/training.rst:297]:
```suggestion
<https://github.com/facebookresearch/fairscale/>`__, so you will have to install this library.
```

File: docs/source/training.rst [hunk 3]
```diff
@@ -278,6 +278,40 @@ pass it to the trainer.
...
+
+This feature requires distributed training.
```
Comment (sgugger) [inline docs/source/training.rst:299]:
```suggestion
This feature requires distributed training (so multiple GPUs) and is not implemented for TPUs.
```

File: docs/source/training.rst [hunk 4]
```diff
@@ -278,6 +278,40 @@ pass it to the trainer.
...
+
+For example here is how you could use it for `finetune_trainer.py`:
+
+.. code-block:: bash
+
+    cd examples/seq2seq
+    python -m torch.distributed.launch --nproc_per_node=2 ./finetune_trainer.py --model_name_or_path sshleifer/distill-mbart-en-ro-12-4 --output_dir output_dir --adam_eps 1e-06 --data_dir wmt_en_ro --do_train --freeze_embeds --label_smoothing 0.1 --learning_rate 3e-5 --logging_first_step --logging_steps 1000 --max_source_length 128 --max_target_length 128 --num_train_epochs 1 --overwrite_output_dir --per_device_train_batch_size 4 --sortish_sampler --src_lang en_XX --task translation --tgt_lang ro_RO  --warmup_steps 500 --n_train 500 --fp16 --sharded_ddp
```
Comment (sgugger) [inline docs/source/training.rst:306]: Can we format the example on sevarl lines using `\`? Also I think we can remove lots of options to make the example simpler: label smoothing, using the default learning rate (which is 5e-5 so shouldn't change much)?

File: docs/source/training.rst [hunk 5]
```diff
@@ -278,6 +278,40 @@ pass it to the trainer.
...
+
+Note that it works with `--fp16` too, to make things even faster.
+
+One of the main benefits of enabling `--sharded_ddp` is that you should be able to use significantly larger batch sizes
+using the same hardware (e.g. 3x or bigger).
```
Comment (sgugger) [inline docs/source/training.rst:311]:
```suggestion
One of the main benefits of enabling `--sharded_ddp` is that it uses a lot less GPU memory, so you should be able to
use significantly larger batch sizes using the same hardware (e.g. 3x or bigger).
```

File: docs/source/training.rst [hunk 6]
```diff
@@ -278,6 +278,40 @@ pass it to the trainer.
...
+
+Eventually more parts will be supported via integrating `DeepSpeed <https://github.com/microsoft/DeepSpeed>`_.
```
Comment (sgugger) [inline docs/source/training.rst:313]:
```suggestion
Eventually more parts will be supported via integrating `DeepSpeed <https://github.com/microsoft/DeepSpeed>`__.
```

**Review by sgugger — APPROVED** [review]
Thanks for drafting this!

File: docs/source/training.rst (hunk 4 above)
Comment (stas00) [reply docs/source/training.rst]: Good idea - I removed even more of them.

File: docs/source/training.rst [hunk 7]
```diff
@@ -278,6 +278,46 @@ pass it to the trainer.
...
+    --per_device_train_batch_size 1  --freeze_embeds \
+    --src_lang en_XX --tgt_lang ro_RO --task translation \
+    --fp16 --sharded_ddp
+
+Note that it works with `--fp16` too, to make things even faster.
+
+One of the main benefits of enabling `--sharded_ddp` is that it uses a lot less GPU memory, so you should be able to
+use significantly larger batch sizes using the same hardware (e.g. 3x or bigger).
+
+Eventually more parts will be supported via integrating `DeepSpeed <https://github.com/microsoft/DeepSpeed>`__.
```
Comment (blefaudeux) [inline docs/source/training.rst:319]: just to be clear, fairscale and deepspeed do not share code, so it's not a natural follow up (by no way against that, feel free of course, it's just that somebody reading this could understand it that way). Fairscale/OSS & shardedDDP are certainly based on the ZeRO paper ideas and the credit here is very valid to me, I'm not disputing that of course.

Comment (sgugger) [reply docs/source/training.rst:319]: We can make it clearer that they are both different implementations of the ZeRO paper.

Comment (stas00) [reply docs/source/training.rst:319]: @blefaudeux, we want to make sure that:

1. both fairscale and deepspace will get the full awesomeness factor - so please do make any suggestions that you see fit - both of your projects are amazing!
2. the users have an easy way to understand when to use which and how they can evaluate pros and cons - so again any suggestions for clarification are very welcome.

The deepspeed integration PR https://github.com/huggingface/transformers/pull/9211  is coming along nicely so we will expand and make this doc section even more clear and balanced - this was just the very basic entry point to send users to when they ask - what's `sharded_ddp` is about, where to read about it, what nuances to know about, etc.
